pyro/distributions/transforms/matrix_exponential.py

```diff
@@ -0,0 +1,174 @@
+        """
+
+        M = self.weights
+        return self._trace(M)
+
+
+@copy_docs_from(ConditionedMatrixExponential)
+class MatrixExponential(ConditionedMatrixExponential, TransformModule):
+    r"""
+    A dense matrix exponential bijective transform (Hoogeboom et al., 2020) with
+    equation,
+
+        :math:`\mathbf{y} = \exp(M)\mathbf{x}`
+
+    where :math:`\mathbf{x}` are the inputs, :math:`\mathbf{y}` are the outputs,
+    :math:`\exp(\cdot)` represents the matrix exponential, and the learnable
+    parameters are :math:`M\in\mathbb{R}^D\times\mathbb{R}^D` for input dimension
+    :math:`D`. In general, :math:`M` is not required to be invertible.
+
+    Due to the favourable mathematical properties of the matrix exponential, the
+    transform has an exact inverse and a log-determinate-Jacobian that scales in
+    time-complexity as :math:`O(D)`. Both the forward and reverse operations are
+    approximated with a truncated power series. For numerical stability, the
+    spectral norm of :math:`M` is restricted.
+
+    Example usage:
+
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> transform = MatrixExponential(10)
+    >>> pyro.module("my_transform", transform)  # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [transform])
+    >>> flow_dist.sample()  # doctest: +SKIP
+
+    :param input_dim: the dimension of the input (and output) variable.
+    :type input_dim: int
+    :param iterations: the number of terms to use in the truncated power series that
+        approximates matrix exponentiation.
+    :type iterations: int
+    :param normalization: One of `['none', 'weight', 'spectral']` normalization that
+        selects what type of normalization to apply to the weight matrix. `weight`
+        corresponds to weight normalization (Salimans and Kingma, 2016) and
+        `spectral` to spectral normalization (Miyato et al, 2018).
+    :type normalization: string
+    :param bound: a bound on either the weight or spectral norm, when either of
+        those two types of regularization are chosen by the `normalization`
+        argument. A lower value for this results in fewer required terms of the
+        truncated power series to closely approximate the exact value of the matrix
+        exponential.
+    :type bound: float
+
+    References:
+
+    [1] Emiel Hoogeboom, Victor Garcia Satorras, Jakub M. Tomczak, Max Welling. The
+        Convolution Exponential and Generalized Sylvester Flows. [arXiv:2006.01910]
+    [2] Tim Salimans, Diederik P. Kingma. Weight Normalization: A Simple
+        Reparameterization to Accelerate Training of Deep Neural Networks.
+        [arXiv:1602.07868]
+    [3] Takeru Miyato, Toshiki Kataoka, Masanori Koyama, Yuichi Yoshida. Spectral
+        Normalization for Generative Adversarial Networks. ICLR 2018.
+
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
```

Comment:
nit: I guess we could use `constraints.real_vector` here. Do you know if that breaks anything?

Reply:
It doesn't seem to break anything as far as I'm aware... I should make this change then for the other `event_dim == 1` transforms

Reply:
There's no `constraints.real_tensor` though for when `event_dim > 1`...

Reply:
I've made these changes for all transforms except for `GeneralizedChannelPermute`